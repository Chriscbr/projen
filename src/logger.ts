import * as chalk from 'chalk';
import { Component } from './component';
import { Project } from './project';

export class Logger extends Component {
  private _level: LogLevel;
  private usePrefix: boolean;

  constructor(project: Project) {
    super(project);

    this._level = LogLevel.INFO;
    this.usePrefix = false;
  }

  public get level() {
    return this._level;
  }

  public set level(level: LogLevel) {
    this._level = level;
  }

  public log(level: LogLevel, ...text: any[]) {
    if (level === LogLevel.OFF) {
      throw new Error('Cannot log a message with level LogLevel.OFF');
    }
    if (level <= this.level) {
      const color = this.colorForLogLevel(level);
      const prefix = this.usePrefix ? `[${this.project.name}] ` : '';
      console.error(`ðŸ¤– ${prefix}${color(...text)}`);
    }
  }

  public verbose(...text: any[]) {
    this.log(LogLevel.VERBOSE, ...text);
  }

  public debug(...text: any[]) {
    this.log(LogLevel.DEBUG, ...text);
  }

  public info(...text: any[]) {
    this.log(LogLevel.INFO, ...text);
  }

  public warn(...text: any[]) {
    this.log(LogLevel.WARN, ...text);
  }

  public error(...text: any[]) {
    this.log(LogLevel.ERROR, ...text);
  }

  private colorForLogLevel(level: LogLevel): chalk.ChalkFunction {
    switch (level) {
      case LogLevel.ERROR: return chalk.red;
      case LogLevel.WARN: return chalk.yellow;
      case LogLevel.INFO: return chalk.cyan;
      case LogLevel.DEBUG: return chalk.gray;
      case LogLevel.VERBOSE: return chalk.white;
      default:
        return chalk.white;
    }
  }
}

export enum LogLevel {
  OFF = '00.off',
  ERROR = '10.error',
  WARN = '20.warn',
  INFO = '30.info',
  DEBUG = '40.debug',
  VERBOSE = '50.verbose'
}