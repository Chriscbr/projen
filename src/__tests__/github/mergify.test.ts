import { AutoMerge, Mergify } from '../../github';
import { LogLevel } from '../../logger';
import { NodeProject, NodeProjectOptions } from '../../node-project';
import { mkdtemp, synthSnapshot } from '../util';

describe('mergify', () => {
  test('default', () => {
    // GIVEN
    const project = createProject();

    // WHEN
    // THEN
    const snapshot = synthSnapshot(project);
    expect(snapshot['.mergify.yml']).toBeDefined();
    expect(snapshot['.mergify.yml']).toMatchSnapshot();
  });

  describe('manually enabled with auto-merge', () => {
    test('default', () => {
      // GIVEN
      const project = createProject({ mergify: false });

      // WHEN
      const mergify = new Mergify(project.github!);
      new AutoMerge(project, { mergify: mergify });

      // THEN
      const snapshot = synthSnapshot(project);
      expect(snapshot['.mergify.yml']).toBeDefined();
      expect(snapshot['.mergify.yml']).toMatchSnapshot();
    });

    test('with options', () => {
      // GIVEN
      const project = createProject({ mergify: false });

      // WHEN
      const mergify = new Mergify(project.github!);
      new AutoMerge(project, {
        mergify: mergify,
        withoutLabels: ['do-not-merge', 'missing-tests'],
        approvedReviews: 3,
      });

      // THEN
      const snapshot = synthSnapshot(project);
      expect(snapshot['.mergify.yml']).toBeDefined();
      expect(snapshot['.mergify.yml']).toContain('- -label~=(do-not-merge|missing-tests)');
      expect(snapshot['.mergify.yml']).toContain('- "#approved-reviews-by>=3"');
      expect(snapshot['.mergify.yml']).toMatchSnapshot();
    });
  });
});

type ProjectOptions = Omit<NodeProjectOptions, 'outdir' | 'defaultReleaseBranch' | 'name'>;
function createProject(options: ProjectOptions = {}): NodeProject {
  return new NodeProject({
    outdir: mkdtemp(),
    defaultReleaseBranch: 'main',
    name: 'node-project',
    logging: {
      level: LogLevel.OFF,
    },
    ...options,
  });
}