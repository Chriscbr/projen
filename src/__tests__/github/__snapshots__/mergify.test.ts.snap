// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergify default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      merge:
        method: squash
        commit_message: title+body
        strict: smart
        strict_method: merge
      delete_head_branch: {}
    conditions:
      - \\"#approved-reviews-by>=1\\"
      - status-success=build
"
`;

exports[`mergify manually enabled with auto-merge default 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      merge:
        method: squash
        commit_message: title+body
        strict: smart
        strict_method: merge
      delete_head_branch: {}
    conditions:
      - \\"#approved-reviews-by>=1\\"
"
`;

exports[`mergify manually enabled with auto-merge with options 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      merge:
        method: squash
        commit_message: title+body
        strict: smart
        strict_method: merge
      delete_head_branch: {}
    conditions:
      - \\"#approved-reviews-by>=3\\"
      - -label~=(do-not-merge|missing-tests)
"
`;
